presto-parquet/src/main/java/com/facebook/presto/parquet/batchreader/decoders/plain/ShortDecimalFixedWidthByteArrayBatchDecoder.java

```diff
@@ -0,0 +1,274 @@
+            implements ShortDecimalDecoder
+    {
+        @Override
+        public void decode(SimpleSliceInputStream input, long[] values, int offset, int length)
+        {
+            int bytesOffSet = 0;
+            int endOffset = offset + length;
+            int i = offset;
+            for (; i < endOffset - 2; i += 2) {
+                // We read redundant bytes and then ignore them. Sign bit is propagated by `>>` operator
+                long value = Long.reverseBytes(input.getLongUnsafe(bytesOffSet));
+                values[i] = value >> 40;
+                values[i + 1] = value << 24 >> 40;
+                bytesOffSet += 6;
+            }
+            // Decode the last values "normally" as it would read data out of bounds
+            while (i < endOffset) {
+                values[i++] = decode(input, bytesOffSet);
+                bytesOffSet += 3;
+            }
+            input.skip(bytesOffSet);
+        }
+
+        private long decode(SimpleSliceInputStream input, int index)
+        {
+            long value = (input.getByteUnsafe(index + 2) & 0xFFL)
+                    | (input.getByteUnsafe(index + 1) & 0xFFL) << 8
+                    | (input.getByteUnsafe(index) & 0xFFL) << 16;
+            return propagateSignBit(value, 40);
+        }
+    }
+
+    private static final class BigEndianReader2
+            implements ShortDecimalDecoder
+    {
+        @Override
+        public void decode(SimpleSliceInputStream input, long[] values, int offset, int length)
+        {
+            while (length > 3) {
+                long value = input.readLongUnsafe();
+                // Reverse all bytes at once
+                value = Long.reverseBytes(value);
+
+                // We first shift the byte as left as possible. Then, when shifting back right,
+                // the sign bit will get propagated
+                values[offset] = value >> 48;
+                values[offset + 1] = value << 16 >> 48;
+                values[offset + 2] = value << 32 >> 48;
+                values[offset + 3] = value << 48 >> 48;
+
+                offset += 4;
+                length -= 4;
+            }
+
+            while (length > 0) {
+                // Implicit cast will propagate the sign bit correctly, as it is performed after the byte reversal.
+                values[offset++] = Short.reverseBytes(input.readShort());
+                length--;
+            }
+        }
+    }
+
+    private static final class BigEndianReader1
+            implements ShortDecimalDecoder
+    {
+        @Override
+        public void decode(SimpleSliceInputStream input, long[] values, int offset, int length)
+        {
+            while (length > 7) {
+                long value = input.readLongUnsafe();
+
+                // We first shift the byte as left as possible. Then, when shifting back right,
+                // the sign bit will get propagated
+                values[offset] = value << 56 >> 56;
```

Comment:
Order of operations is foggy. Please use parentheses to make this explicit.

Reply:
Fixed.